_posts/2020-02-05-keycloak-z-angularem-i-keycloakiem.md

```diff
@@ -0,0 +1,427 @@
+        }
+    }
+
+    isAccessAllowed(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
+        return new Promise(async (resolve) => {
+            if (!this.authenticated) {
+                return this.keycloakService.login();
+            }
+            resolve(this.roles.includes(this.requiredUserRole));
+        });
+    }
+}
+```
+
+a następnie wykorzystamy naszego guarda w `app-routing.module.ts` do zabezpieczenia routa `protected`:
+
+```typescript
+const routes: Routes = [
+    {
+        path: '',
+        component: PublicComponent
+    },
+    {
+        path: 'protected',
+        component: ProtectedComponent,
+        canActivate: [AppAuthGuard]
+    }
+];
+```
+
+## Testy
+
+### Uruchomienie
+
+Jeśli mamy już uruchomiony serwer Keycloak, możemy zbudować nasze aplikacje i obraz dockerowy, a następnie uruchomić kontener.
+
+W tym celu w głównym folderze repozytorium najpierw budujemy aplikacje:
+
+```shell
+./gradlew build
+```
+
+po zbudowaniu tworzymy obraz dockerowy:
+
+```shell
+docker build -t michuu93/spring-angular-keycloak-demo .
+```
+
+a na koniec uruchamiamy kontener:
+
+```shell
+docker run --rm --name spring-angular-keycloak-demo --network host michuu93/spring-angular-keycloak-demo
+```
+
+Ponieważ zarówno nasze aplikacje z wnętrza kontenera, jak i my z przeglądarki internetowej musimy mieć dostęp do serwera Keycloak pod tym samym adresem, nie wnikając w zasadę działania sieci w dockerze uruchomiliśmy kontener z opcją `--network host`.
+
+Dzięki temu nie musimy wystawiać dodatkowo żadnego proxy ani podmieniać hostów w naszym systemie. Jest to jednak ułatwienie na potrzeby demo i nie powinniśmy go wykorzystywać na produkcji.
+
+### Weryfikacja działania uwierzytelniania
+
+Po uruchomieniu możemy sprawdzić nasze aplikacje w działaniu!
+
+Przechodzimy na stronę [localhost:9082](http://localhost:9082/) i powinniśmy zobaczyć aplikację frontendową, a dokładniej nasz `ToolbarComponent` oraz `PublicComponent`:
+
+![Public component](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/public_component_not_logged.png)
+
+Do dyspozycji mamy menu, z którego możemy przejść do:
+
+- `Public` - niezabezpieczonego route z `PublicComponent`
+- `Protected` - zabezpieczonego route z `ProtectedComponent`
+- `Keycloak Configuration` - niezabezpieczonego endpointu udostępniającego konfiguację Keycloak z backendu
+- `Backend Hello` - zabezpieczonego endpointu zwracającego tekst `Hello from the Backend!`
+- `Logout (backend)` - wylogowania z backendu
+- `Logout (frontend)` - przycisk widoczny tylko po zalogowaniu na frontendzie, który wylogowuje nas z frontendu
+
+Sprawdźmy więc czy mamy dostęp do `Keycloak Configuration` bez zalogowania:
+![Keycloak Configuration](/assets/img/posts/2020-02-05-keycloak-z-angularem-i-keycloakiem/keycloak_config.png)
+
+Wygląda na to, że nasz publicznie dostępny route i endpoint api działa poprawnie.
```

Comment:
przecinek + liczba mnoga
Wygląda na to, że nasz publicznie dostępny route i endpoint api, działają poprawnie.